ekotrace-capi/ekotrace-capi-impl/src/lib.rs

```diff
@@ -286,20 +305,151 @@ pub unsafe fn ekotrace_merge_fixed_size_snapshot(
-            }
+            };
         }
     };
     tracer.now()
 }
 
+/// Prepare to write a chunked report.
+///
+/// Populates the out-parameter `out_report_token` with
+/// a value that will be used to produce the
+/// chunks for the report in calls to
+/// `ekotrace_write_next_report_chunk` and `ekotrace_finish_chunked_report`
+///
+/// Once this method has been called, mutating operations on
+/// the Ekotrace instance will return `EKOTRACE_RESULT_REPORT_LOCK_CONFLICT_ERROR`
+/// until all available chunks have been written with  `ekotrace_write_next_report_chunk`
+/// and `ekotrace_finish_chunked_report` called.
+///
+/// # Safety
+///
+/// The Ekotrace instance pointer must be non-null and point
+/// to an initialized instance operating in a single-threaded
+/// fashion.
+///
+/// If the function returns any result besides EKOTRACE_RESULT_OK,
+/// there has been an error, and the out-pointer for the chunked report token
+/// will not be populated.
+#[cfg_attr(feature = "no_mangle", no_mangle)]
+pub unsafe fn ekotrace_start_chunked_report(
+    tracer: *mut Ekotrace<'static>,
+    out_report_token: *mut ChunkedReportToken,
+) -> EkotraceResult {
+    let tracer = match tracer.as_mut() {
+        Some(t) => t,
+        None => return EKOTRACE_RESULT_NULL_POINTER,
+    };
+    match tracer.start_chunked_report() {
+        Ok(token) => {
+            *out_report_token = token;
+            EKOTRACE_RESULT_OK
+        }
+        Err(e) => chunked_report_error_to_ekotrace_result(e),
+    }
+}
+
+fn chunked_report_error_to_ekotrace_result(cre: ChunkedReportError) -> EkotraceResult {
+    match cre {
+        ChunkedReportError::ReportError(re) => report_error_to_ekotrace_result(re),
+        ChunkedReportError::NoChunkedReportInProgress => {
+            EKOTRACE_RESULT_NO_CHUNKED_REPORT_IN_PROGRESS
+        }
+    }
+}
+/// Write up to 1 chunk of a report into
+/// the provided destination buffer.
+///
+/// Populates the out-parameter `out_written_bytes` with
+/// the number of report bytes written into the destination.
+///
+/// If the `out_written_bytes` == 0, then no chunk was
+/// written and there are no chunks left in the report.
+///
+/// The provided ChunkedReportToken should match the value
+/// populated by the `ekotrace_start_chunked_report` call
+/// at the start of this chunked report.
+///
+/// # Safety
+///
+/// The Ekotrace instance pointer must be non-null and point
+/// to an initialized instance operating in a single-threaded
+/// fashion.
+///
+/// The destination buffer pointer must be non-null
+/// and should have an associated size == 256 bytes.
+///
+/// If the function returns any result besides EKOTRACE_RESULT_OK,
+/// there has been an error and the contents of the report destination
+/// are not certain to be initialized or a valid chunk.
+#[cfg_attr(feature = "no_mangle", no_mangle)]
+pub unsafe fn ekotrace_write_next_report_chunk(
+    tracer: *mut Ekotrace<'static>,
+    report_token: *const ChunkedReportToken,
```

Comment:
I'm a little surprised to see the report token being passed around by reference, rather than by value. Any particular reason?

Reply:
In rust-land, the token is opaque and created on report-start. Then it can be used by reference for each chunk-write without being duplicated or introspected, and then consumed at report-finish.

Reply:
The by-default disallowance of token copying in Rust-land is a protective measure to reduce the odds of tokens / reporting sessions getting their wires crossed, so to speak.